In [37]:
import pandas as pd

**1. Introduction**

In [38]:
'''
Main analysis
'''

df = pd.read_csv('Amazon_Sale_Report.csv', index_col='index')
print(f"Rows, columns: {df.shape} ")
df.head(n=3)

Rows, columns: (128975, 23) 


/tmp/ipython-input-383/1608414725.py:5: DtypeWarning: Columns (23) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('Amazon_Sale_Report.csv', index_col='index')


Order ID      Date                        Status Fulfilment  \
index                                                                           
0      405-8078784-5731545  04-30-22                     Cancelled   Merchant   
1      171-9198151-1101146  04-30-22  Shipped - Delivered to Buyer   Merchant   
2      404-0687676-7273146  04-30-22                       Shipped     Amazon   

      Sales Channel  ship-service-level    Style              SKU Category  \
index                                                                        
0          Amazon.in           Standard   SET389   SET389-KR-NP-S      Set   
1          Amazon.in           Standard  JNE3781  JNE3781-KR-XXXL    kurta   
2          Amazon.in          Expedited  JNE3371    JNE3371-KR-XL    kurta   

      Size  ... currency  Amount    ship-city   ship-state  ship-postal-code  \
index       ...                                                                
0        S  ...      INR  647.62       MUMBAI  MAHARASHTRA          400081.0   
1      3XL  ...      INR  406.00    BENGALURU    KARNATAKA          560085.0   
2       XL  ...      INR  329.00  NAVI MUMBAI  MAHARASHTRA          410210.0   

      ship-country                                      promotion-ids    B2B  \
index                                                                          
0               IN                                                NaN  False   
1               IN  Amazon PLCC Free-Financing Universal Merchant ...  False   
2               IN       IN Core Free Shipping 2015/04/08 23-48-5-108   True   

      fulfilled-by Unnamed: 22  
index                           
0        Easy Ship         NaN  
1        Easy Ship         NaN  
2              NaN         NaN  

[3 rows x 23 columns]

**2. Data Overview**

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 128975 entries, 0 to 128974
Data columns (total 23 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Order ID            128975 non-null  object 
 1   Date                128975 non-null  object 
 2   Status              128975 non-null  object 
 3   Fulfilment          128975 non-null  object 
 4   Sales Channel       128975 non-null  object 
 5   ship-service-level  128975 non-null  object 
 6   Style               128975 non-null  object 
 7   SKU                 128975 non-null  object 
 8   Category            128975 non-null  object 
 9   Size                128975 non-null  object 
 10  ASIN                128975 non-null  object 
 11  Courier Status      122103 non-null  object 
 12  Qty                 128975 non-null  int64  
 13  currency            121180 non-null  object 
 14  Amount              121180 non-null  float64
 15  ship-city           128942 non-null  ob

In [40]:
df.describe()

Qty         Amount  ship-postal-code
count  128975.000000  121180.000000     128942.000000
mean        0.904431     648.561465     463966.236509
std         0.313354     281.211687     191476.764941
min         0.000000       0.000000     110001.000000
25%         1.000000     449.000000     382421.000000
50%         1.000000     605.000000     500033.000000
75%         1.000000     788.000000     600024.000000
max        15.000000    5584.000000     989898.000000

As we can see, dataset contains 128975 rows and 23 columns. Several columns have missing values - for example column 'Unnamed' is very suspicious. Let's make a copy of our dataset to ensure that original data will stay untouched.

In [41]:
df_copy = df.copy()

**3. Missing values analysis**

In [72]:
# missing values all width - full na row

df_copy.isna().all()

order id              False
date                  False
status                False
fulfilment            False
sales channel         False
ship-service-level    False
style                 False
sku                   False
category              False
size                  False
asin                  False
courier status        False
qty                   False
currency              False
amount                False
ship-city             False
ship-state            False
ship-postal-code      False
ship-country          False
b2b                   False
dtype: bool

The first column to further examination is 'Unnamed' column.

In [42]:
df_copy['Unnamed: 22']

index
0           NaN
1           NaN
2           NaN
3           NaN
4           NaN
          ...  
128970    False
128971    False
128972    False
128973    False
128974    False
Name: Unnamed: 22, Length: 128975, dtype: object

In [43]:
unnamed_counts = df_copy['Unnamed: 22'].value_counts()
print(f"Values in 'Unnamed 22' column: {unnamed_counts}")
print("")
unnamed_missing = df_copy['Unnamed: 22'].isna().sum()
print(f"Values missing in column 'Unnamed 22': {unnamed_missing}")

Values in 'Unnamed 22' column: Unnamed: 22
False    79925
Name: count, dtype: int64

Values missing in column 'Unnamed 22': 49050


As we can see - this column is not truly informative.  <br>
✅ - TO DO: Column 'Unnamed 22' is to be deleted.

In [44]:
df_copy.drop('Unnamed: 22', axis=1, inplace=True)
df_copy.head(3)

Order ID      Date                        Status Fulfilment  \
index                                                                           
0      405-8078784-5731545  04-30-22                     Cancelled   Merchant   
1      171-9198151-1101146  04-30-22  Shipped - Delivered to Buyer   Merchant   
2      404-0687676-7273146  04-30-22                       Shipped     Amazon   

      Sales Channel  ship-service-level    Style              SKU Category  \
index                                                                        
0          Amazon.in           Standard   SET389   SET389-KR-NP-S      Set   
1          Amazon.in           Standard  JNE3781  JNE3781-KR-XXXL    kurta   
2          Amazon.in          Expedited  JNE3371    JNE3371-KR-XL    kurta   

      Size  ... Qty currency  Amount    ship-city   ship-state  \
index       ...                                                  
0        S  ...   0      INR  647.62       MUMBAI  MAHARASHTRA   
1      3XL  ...   1      INR  406.00    BENGALURU    KARNATAKA   
2       XL  ...   1      INR  329.00  NAVI MUMBAI  MAHARASHTRA   

      ship-postal-code ship-country  \
index                                 
0             400081.0           IN   
1             560085.0           IN   
2             410210.0           IN   

                                           promotion-ids    B2B fulfilled-by  
index                                                                         
0                                                    NaN  False    Easy Ship  
1      Amazon PLCC Free-Financing Universal Merchant ...  False    Easy Ship  
2           IN Core Free Shipping 2015/04/08 23-48-5-108   True          NaN  

[3 rows x 22 columns]

Let's check other missing values.

In [45]:
df_copy.isnull().sum().sort_values()
missing_values_pr = df_copy.isnull().sum() / len(df_copy) * 100
missing_values_pr.sort_values()

Order ID               0.000000
Date                   0.000000
Status                 0.000000
Fulfilment             0.000000
Sales Channel          0.000000
ship-service-level     0.000000
Style                  0.000000
SKU                    0.000000
Category               0.000000
Size                   0.000000
ASIN                   0.000000
Qty                    0.000000
B2B                    0.000000
ship-city              0.025586
ship-state             0.025586
ship-country           0.025586
ship-postal-code       0.025586
Courier Status         5.328164
Amount                 6.043807
currency               6.043807
promotion-ids         38.110487
fulfilled-by          69.546811
dtype: float64

As we can see - other columns to further examination are: fulfilled-by, promotion-ids, currency, Amount. But first - let's change all column names to lower case. <br>
✅ - TO DO: change column names to lower

In [46]:
df_copy.columns = df_copy.columns.str.lower()
df_copy.columns


Index(['order id', 'date', 'status', 'fulfilment', 'sales channel ',
       'ship-service-level', 'style', 'sku', 'category', 'size', 'asin',
       'courier status', 'qty', 'currency', 'amount', 'ship-city',
       'ship-state', 'ship-postal-code', 'ship-country', 'promotion-ids',
       'b2b', 'fulfilled-by'],
      dtype='object')

Since column 'fulfilled-by' is missing 70% of values and column 'promotion-ids' is missing 40% - I've decided to drop such columns.  <br>
✅ - TO DO : drop 'fulfilled_by', drop: 'promotion-ids'

In [47]:
df_copy.drop('fulfilled-by', axis=1, inplace=True)
df_copy.drop('promotion-ids',axis=1,inplace=True)
df_copy.head(3)

order id      date                        status fulfilment  \
index                                                                           
0      405-8078784-5731545  04-30-22                     Cancelled   Merchant   
1      171-9198151-1101146  04-30-22  Shipped - Delivered to Buyer   Merchant   
2      404-0687676-7273146  04-30-22                       Shipped     Amazon   

      sales channel  ship-service-level    style              sku category  \
index                                                                        
0          Amazon.in           Standard   SET389   SET389-KR-NP-S      Set   
1          Amazon.in           Standard  JNE3781  JNE3781-KR-XXXL    kurta   
2          Amazon.in          Expedited  JNE3371    JNE3371-KR-XL    kurta   

      size        asin courier status  qty currency  amount    ship-city  \
index                                                                      
0        S  B09KXVBD7Z            NaN    0      INR  647.62       MUMBAI   
1      3XL  B09K3WFS32        Shipped    1      INR  406.00    BENGALURU   
2       XL  B07WV4JV4D        Shipped    1      INR  329.00  NAVI MUMBAI   

        ship-state  ship-postal-code ship-country    b2b  
index                                                     
0      MAHARASHTRA          400081.0           IN  False  
1        KARNATAKA          560085.0           IN  False  
2      MAHARASHTRA          410210.0           IN   True

Further 'currency' column analysis

In [50]:
df_copy['currency'].value_counts()
df_copy['currency']

index
0         INR
1         INR
2         INR
3         INR
4         INR
         ... 
128970    INR
128971    INR
128972    INR
128973    INR
128974    INR
Name: currency, Length: 128975, dtype: object

As we can see - 'INR' is the only value in such column. I've decided to fill null values using 'INR'. <br>
- ✅ TO DO : Fill null values in 'currency' column with 'INR'

In [55]:
df_copy['currency'] = df_copy['currency'].fillna('INR')

In [56]:
# check
df_copy['currency'].isna().sum()

np.int64(0)

Further 'amount' column analysis

In [60]:
df_copy['amount']

index
0          647.62
1          406.00
2          329.00
3          753.33
4          574.00
           ...   
128970     517.00
128971     999.00
128972     690.00
128973    1199.00
128974     696.00
Name: amount, Length: 128975, dtype: float64

Since 'amount' colum contains float values - I will fill null values using mean value.  <br>
- ✅ TO DO: fill null 'amount' values using mean

In [65]:
mean_val = df_copy['amount'].mean().round(2)
df_copy['amount'] = df_copy['amount'].fillna(mean_val)

In [66]:
#check
df_copy['amount'].isna().sum()

np.int64(0)

Further 'courier status' column analysis

In [69]:
df_copy['courier status'].value_counts()

courier status
Shipped      109487
Unshipped      6681
Cancelled      5935
Name: count, dtype: int64

For the courier_status column, we considered two imputation strategies: filling missing values with the mode ('Shipped') or labeling them as 'Unknown'. I've decided to proceed with the latter to maintain data integrity and avoid making assumptions about the logistics process.
- ✅ TO DO: Fill missing 'courier status' values with 'Unknown'

In [74]:
df_copy['courier status'] = df_copy['courier status'].fillna('Unknown')

In [75]:
# check
df_copy['courier status'].isna().sum()

np.int64(0)

Now, let's handle missing values in columns:
'ship-city', 'ship-state', 'ship-country','ship-postal-code'

In [78]:
df_copy['ship-city'].value_counts()

ship-city
BENGALURU              11217
HYDERABAD               8074
MUMBAI                  6126
NEW DELHI               5795
CHENNAI                 5421
                       ...  
shimoga                    1
Bangluru                   1
Dombiwali                  1
Begumpet, Hyderabad        1
Badlapur , Thane           1
Name: count, Length: 8955, dtype: int64

In [80]:
df_copy['ship-state'].value_counts()

ship-state
MAHARASHTRA      22260
KARNATAKA        17326
TAMIL NADU       11483
TELANGANA        11330
UTTAR PRADESH    10638
                 ...  
Nagaland             1
Mizoram              1
bihar                1
rajsthan             1
Rajsthan             1
Name: count, Length: 69, dtype: int64

Since both ship-city and ship-state exhibit high cardinality, filling missing values with the most frequent category (mode) would significantly distort the geographical distribution. Therefore, we have opted to impute all NaNs with 'Unknown' to maintain the accuracy of our location-based analysis <br>
✅ TO DO
- fill null values in columns 'ship-city', 'ship-state' with 'Unknown'

In [85]:
df_copy[['ship-city','ship-state']]=  df_copy[['ship-city', 'ship-state']].fillna('Unknown')

In [86]:
#check
df_copy[['ship-city', 'ship-state']].isna().sum()

ship-city     0
ship-state    0
dtype: int64

In [83]:
df_copy['ship-postal-code']

index
0         400081.0
1         560085.0
2         410210.0
3         605008.0
4         600073.0
            ...   
128970    500013.0
128971    122004.0
128972    500049.0
128973    389350.0
128974    492014.0
Name: ship-postal-code, Length: 128975, dtype: float64

During the analysis, we identified that the postal-code column was stored as a float. To ensure data consistency, we converted it to integer format. Regarding missing values, since postal codes are highly specific, we filled them with a placeholder value (0) to distinguish them from valid codes while maintaining a numeric data type <br>
✅ TO DO: retype 'postal-code' to 'int' and fill missing values with 0

In [90]:
# We can not convert values before dealing with null-values.
df_copy['ship-postal-code'] = df_copy['ship-postal-code'].fillna(0.0)

df_copy['ship-postal-code']= df_copy['ship-postal-code'].astype('int')

In [91]:
#check
df_copy['ship-postal-code'].isna().sum()

np.int64(0)

In [82]:
df_copy['ship-country']

index
0         IN
1         IN
2         IN
3         IN
4         IN
          ..
128970    IN
128971    IN
128972    IN
128973    IN
128974    IN
Name: ship-country, Length: 128975, dtype: object

Since in 'ship-country' we only have IN & Null values I've decided to fill missing values with 'IN' <br>
✅ TO DO: Fill missing values in 'ship-country' with 'IN'

In [92]:
df_copy['ship-country'] = df_copy['ship-country'].fillna('IN')

In [93]:
#check
df_copy['ship-country'].isna().sum()

np.int64(0)

After this step, out dataset should be null-less :)

In [94]:
df_copy.isna().sum()

order id              0
date                  0
status                0
fulfilment            0
sales channel         0
ship-service-level    0
style                 0
sku                   0
category              0
size                  0
asin                  0
courier status        0
qty                   0
currency              0
amount                0
ship-city             0
ship-state            0
ship-postal-code      0
ship-country          0
b2b                   0
dtype: int64

Summary: <br>
✅ - TO DO: Column 'Unnamed 22' is to be deleted. <br>
✅ - TO DO: change column names to lower <br>
✅ - TO DO : drop 'fulfilled_by', drop: 'promotion-ids' <br>
✅ TO DO : Fill null values in 'currency' column with 'INR' <br>
✅ TO DO: fill null 'amount' values using mean <br>
✅ TO DO: Fill missing 'courier status' values with 'Unknown' <br>
✅ TO DO fill null values in columns 'ship-city', 'ship-state' with 'Unknown' <br>
✅ TO DO: retype 'postal-code' to 'int' and fill missing values with 0 <br>
✅ TO DO: Fill missing values in 'ship-country' with 'IN' <br>
<br>

In [97]:
df_copy.to_csv('part1.csv')